Dispatcher blocks dispatching once unackMsgs reached max limit

rdhabalia opened the pull request:
### Motivation

as discussed at #398 : if client misbehaves where it doesn't ack messages and keep redelivering unack messages then it causes broker to keep large number of `PositionImpl` instances in memory  until broker dispatches those messages. It may cause frequent old-gc and zk-session lost. Therefore, broker requires a safe-guard to control number of dispatched messages which helps to reduce number of `PositionImpl` objects and gc occurrences. With this patch, It shows less gc and stable broker by running same test mentioned into #398.

### Modifications

Broker blocks dispatching of messages on subscription if given subscriber reaches defined unacked message limit. Broker unblocks dispatching once it receives required number of acked messages and it starts dispatching new messages on that subscriber again.

### Result

Subscriber will not receive any new messages once it reaches to unack message limit. Broker can redeliver unacked messages to blocked-subscriber so, subscriber can ack them back and get unblocked to receieve new messages.

nkurihar commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:
```diff
@@ -190,6 +199,9 @@ private void readMoreEntries() {
                     readMoreEntries();
                 }
+            } else if (BLOCKED_DISPATCHER_ON_UNACKMSG_UPDATER.get(this) == TRUE) {
+                log.warn("[{}] Dispatcher read is blocked due to unackMessages {} reached to max {]", name,
```
small comment: {] looks like typo

nkurihar commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:
```diff
@@ -560,5 +572,32 @@ public synchronized void redeliverUnacknowledgedMessages(Consumer consumer, List
+        int unAckedMessages = TOTAL_UNACKED_MESSAGES_UPDATER.addAndGet(this, numberOfMessages);
+        if (unAckedMessages >= maxUnackedMessages
+                && BLOCKED_DISPATCHER_ON_UNACKMSG_UPDATER.compareAndSet(this, FALSE, TRUE)) {
+            log.info("[{}] Dispatcher is blocked due to unackMessages {} reached to max {]", name,
```
{] looks like typo, too.

nkurihar commented on conf/broker.conf:
```diff
@@ -102,6 +102,12 @@ tlsAllowInsecureConnection=false
+# all consumers once, this limit reaches until consumer starts acknowledging messages back and unack count reaches
+# to maxUnackedMessagesPerDispatcher/2. Using a value of 0, is disabling unackedMessage-limit check and dispatcher
+# can dispatch messages without any restriction
+maxUnackedMessagesPerDispatcher=200000
```
It's nice to have the same configuration in `standalone.conf` too.

merlimat replied:
Dispatcher is kind of an internal implementation terminology, what about `maxUnackedMessagesPerSubscription` ? 

Also, does it really needs to be configurable or should we just use `maxUnackedMessagesPerConsumer * 2`  (or 3).

rdhabalia replied:
> Dispatcher is kind of an internal implementation terminology, what about maxUnackedMessagesPerSubscription ?

Yes, I will update with `maxUnackedMessagesPerSubscription`

> Also, does it really needs to be configurable or should we just use 

yes, because if few of the topics are misbehaving then this limit helps broker but if most of the topics in broker misbehave then this number should be lower. For example in the same test if we have 600 topics and each subscription allows 0.2million msgs then broker ends up with 120mln unack msgs in memory which doesn't help and for that this number should be lower. So, it's better to give option to configure this number.

merlimat replied:
> . For example in the same test if we have 600 topics and each subscription allows 0.2million msgs then broker ends up with 120mln unack msgs in memory which doesn't help and for that this number should be lower. So, it's better to give option to configure this number.

Correct. Then would it be feasible to have an efficient way (maybe with relaxed enforcement) to add a per-broker limit?

saandrews replied:
Yes, the objective of [402](https://github.com/yahoo/pulsar/issues/402) is to come up ways to adjust the limits dynamically.

merlimat replied:
That's a different thing. Here we're talking about `pendingAcks` for shared subscriptions. The dispatcher needs a way to know which messages were sent to a particular consumer, to be able, in case the consumer disconnect, to redeliver the messages somewhere else. 

For short term we could just use this per-subscription limit and switch to `ConcurrentLongLongPairHashMap` that it's already implemented `(long,long)  --> (long,long)`. We don't need the 2nd long in the value, but the overhead is not that big. 

That way, even keeping 10 M of pending acks would require only 320 MB of memory and would not cause GC troubles.

rdhabalia replied:
> Then would it be feasible to have an efficient way (maybe with relaxed enforcement) to add a per-broker limit?

But then it raises question:
1. If we have per-broker limit then broker blocks dispatching to all topics once it reaches that limit: So, if only few of the topics are misbehaving then will broker punish other topics as well and blocks dispatching on other topics as well?
2. once, misbehaving-client acks limit/2 msgs -> broker should notify all dispatchers to dispatch and there could be possibility that again misbehave-client made broker to reach that limit and again dispatching will be blocked. 

So, per-broker limit may not be good idea??

rdhabalia replied:
fixed rename to `maxUnackedMessagesPerSubscription`

merlimat replied:
> So, per-broker limit may not help broker??

I am not saying of doing that as part of this patch.. Just that if number of topic with bad behavior increase it could still use a lot of memory. Maybe doing something like that, when the threshold is reached, only block topics that have a lot of pending messages and leave alone topics with normal behavior.

rdhabalia replied:
> when the threshold is reached, only block topics that have a lot of pending messages and leave alone topics with normal behavior.

Yes, that make sense. In that way, we don't have to change `maxUnackedMessagesPerSubscription` and once per-broker threshold reached, those bad-behaving subscription will be blocked.

rdhabalia commented:
@nkurihar addressed comments.

nkurihar commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/persistent/PersistentTopic.java:
```diff
@@ -969,7 +967,12 @@ public void updateRates(NamespaceStats nsStats, NamespaceBundleStats bundleStats
-                    destStatsStream.writePair("unackedMessages", subUnackedMessages);
+                    if(subscription.getDispatcher() instanceof PersistentDispatcherMultipleConsumers) {
+                        PersistentDispatcherMultipleConsumers dispatcher = (PersistentDispatcherMultipleConsumers)subscription.getDispatcher();
+                        destStatsStream.writePair("blockedDispatcherOnUnackedMsgs",  dispatcher.isBlockedDispatcherOnUnackedMsgs());
```
I tried the simple case:
* standalone cluster
* maxUnackedMessagesPerDispatcher=10
* 2  shared consumers which return no ack
* produce 100 messages

Dispatching is blocked after the unack count reaches the limit (10) as expected,
however, `blockedDispatcherOnUnackedMsgs` in the stats result is still "false":
```
$ bin/pulsar-admin persistent stats persistent://sample/standalone/ns1/t1
{
  "msgRateIn" : 0.0,
  "msgThroughputIn" : 0.0,
  "msgRateOut" : 0.0,
  "msgThroughputOut" : 0.0,
  "averageMsgSize" : 0.0,
  "storageSize" : 3013,
  "publishers" : [ ],
  "subscriptions" : {
    "bad" : {
      "msgRateOut" : 0.0,
      "msgThroughputOut" : 0.0,
      "msgRateRedeliver" : 0.0,
      "msgBacklog" : 100,
      "blockedDispatcherOnUnackedMsgs" : false,
      "unackedMessages" : 10,
      "type" : "Shared",
      "msgRateExpired" : 0.0,
      "consumers" : [ {
        "msgRateOut" : 0.0,
        "msgThroughputOut" : 0.0,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "aa6ed",
        "availablePermits" : 995,
        "unackedMessages" : 5,
        "blockedConsumerOnUnackedMsgs" : false,
        "address" : "/172.16.112.250:53894",
        "connectedSince" : "2017-05-11 09:58:36.780+0900"
      }, {
        "msgRateOut" : 0.0,
        "msgThroughputOut" : 0.0,
        "msgRateRedeliver" : 0.0,
        "consumerName" : "3eba4",
        "availablePermits" : 995,
        "unackedMessages" : 5,
        "blockedConsumerOnUnackedMsgs" : false,
        "address" : "/172.16.112.250:53896",
        "connectedSince" : "2017-05-11 09:58:39.206+0900"
      } ]
    }
  },
  "replication" : { }
}
```

rdhabalia replied:
good catch. missed to update `blockedConsumerOnUnackedMsgs` for `getStatus()` api, pushed the [fix](https://github.com/yahoo/pulsar/pull/399/commits/1f291922e2210aa6b4a11c220d8dd37d3ade66b1)

nkurihar replied:
thanks!

rdhabalia commented:
@saandrews addressed the comment.

Also, adding test result mentioned at #398 before and after change :

Before change:
1. GC: ~20sec old gc
![screenshot from without-changes](http://i.imgur.com/1RZnI6F.png)

2. message delivery: 120K
![screenshot from without-changes](http://i.imgur.com/S0S6JT5.png)


After change:
1. GC: 1.5 sec old gc
![screenshot from without-changes](http://i.imgur.com/E1g35Jp.png)

2. Message delivery : 60K
![screenshot from without-changes](http://i.imgur.com/SSv87PZ.png)

nkurihar commented:
LGTM

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java:
```diff
@@ -507,8 +510,8 @@ public void redeliverUnacknowledgedMessages() {
             int totalRedeliveryMessages = 0;
-            for (Integer batchSize : pendingAcks.values()) {
-                totalRedeliveryMessages += batchSize;
+            for (LongPair batchSize : pendingAcks.values()) {
```
Using `values()` needs to create a copy of the entire hashmap. Try to use for-each for iteration

merlimat replied:
maybe use `AtomicInteger` to accumulate since you cannot just add over the same `int` from a lambda

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java:
```diff
@@ -113,7 +115,7 @@ public Consumer(Subscription subscription, SubType subType, long consumerId, int
 
         if (subType == SubType.Shared) {
-            this.pendingAcks = new ConcurrentOpenHashMap<PositionImpl, Integer>(256, 2);
+            this.pendingAcks = new ConcurrentLongLongPairHashMap(256, 2);
```
I think we should change the map to use 1 single segment since it will not have a lot of contention.

`new ConcurrentLongLongPairHashMap(256, 1)`

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java:
```diff
@@ -251,7 +253,7 @@ int updatePermitsAndPendingAcks(final List<Entry> entries) throws PulsarServerEx
             if (pendingAcks != null) {
                 PositionImpl pos = (PositionImpl) entry.getPosition();
-                pendingAcks.put(pos, batchSize);
+                pendingAcks.put(pos.getLedgerId(), pos.getEntryId(), batchSize, 0);
```
One possible improvement here would be to skip the `PositionImpl` all together. Expose `getLedgerId()` and `getEntryId()` on the `EntryImpl` and do like: 

```java
if (pendingAcks != null) {
    EntryImpl e = (EntryImpl) entry;
    pendingAcks.put(e.getLedgerId(), e.getEntryId(), batchSize, 0);
}
```

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java:
```diff
@@ -465,9 +467,9 @@ public int hashCode() {
+        if (pendingAcks.get(position.getLedgerId(), position.getEntryId()) == null) {
             for (Consumer consumer : subscription.getConsumers()) {
-                if (!consumer.equals(this) && consumer.getPendingAcks().get(position) != null) {
+                if (!consumer.equals(this) && consumer.getPendingAcks().get(position.getLedgerId(), position.getEntryId()) != null) {
```
`consumer.getPendingAcks().containsKey(position.getLedgerId(), position.getEntryId())`

merlimat commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/persistent/PersistentDispatcherMultipleConsumers.java:
```diff
@@ -141,8 +141,8 @@ public synchronized void removeConsumer(Consumer consumer) throws BrokerServiceE
-                consumer.getPendingAcks().forEach((pendingMessages, totalMsg) -> {
-                    messagesToReplay.add(pendingMessages);
+                consumer.getPendingAcks().forEach((ledgerId, entryId, batchSize, none) -> {
+                    messagesToReplay.add(PositionImpl.get(ledgerId, entryId));
```
just use the `new PositionImpl()` since this instance would hardly be possible to recycle anyway.

merlimat commented:
@rdhabalia Once this is merged, can you also redo the same test for which you posted the GC graphs above? Just to see that 1.5s pauses effectively go away.

rdhabalia commented:
> Once this is merged, can you also redo the same test for which you posted the GC graphs above? Just to see that 1.5s pauses effectively go away.

Yes, but we still need one more improvement at `PersistentDispatcherMultipleConsumers.messagesToReplay`. We have seen that on redelivery `Consumer` adds all unackMsg positions to `PersistentDispatcherMultipleConsumers.messagesToReplay` and heap-dump shows that size of this list reached to 120k-200k and if delivery is slow then these objects live in memory for some more time and it can also cause gc.

merlimat commented:
> PersistentDispatcherMultipleConsumers.messagesToReplay and heap-dump shows that size of this list reached to 120k-200k and if delivery is slow then these objects live in memory for some more time and it can also cause gc.

I see : 
```java
    private TreeSet<PositionImpl> messagesToReplay;
```

We use a tree-set to redeliver in order, though it shouldn't be strictly necessary. 
I guess we could either: 
 * Use ConcurrentLongPairSet and forget about order
 * Implement a `Queue< Pair<Long, Long> >` that is actually implemented as an array of primitive longs

rdhabalia commented:
> We use a tree-set to redeliver in order, though it shouldn't be strictly necessary.
I guess we could either:

yes, if we think ordering can be ignored then I will change it with `ConcurrentLongPairSet`.

merlimat commented:
> yes, if we think ordering can be ignored then I will change it with ConcurrentLongPairSet.

The only reason to have ordering there is to fill the gaps in the ack-sequence as soon as possible. Now, since we persist the individually deleted messages, I think that it would be less important to do so.

Also, since the number of unacked messages is limited, the out-of-order is still restricted to that amount of messages.

So we should go ahead with `ConcurrentLongPairSet`, except that we don't have that class ;) 
We need to add a new specialization for the `Set<Pair<Long,Long>>` case.

rdhabalia commented:
> So we should go ahead with ConcurrentLongPairSet, except that we don't have that class ;)
We need to add a new specialization for the Set<Pair<Long,Long>> case.

Yes, can we merge this PR and I will create a separate one with this change.

asafm commented on pulsar-broker/src/main/java/com/yahoo/pulsar/broker/service/Consumer.java:
```diff
@@ -546,5 +553,15 @@ public Subscription getSubscription() {
         return subscription;
     }
     
+    private int addAndGetUnAckedMsgs(Consumer consumer, int ackedMessages) {
```
@rdhabalia Why did you call this `ackedMessages` while you're adding (`addUnackedMessages`) - it's very confusing. I had to search through the code to find you added `-` to it 2 calls up the stack. Very time consuming.
